src/query/service/src/pipelines/processors/transforms/hash_join/probe_join/inner_join.rs

```diff
@@ -81,32 +99,55 @@ impl JoinHashTable {
                                 ));
                             }
 
-                            let addition = probe_indexes.capacity() - probe_indexes.len();
+                            let addition = JOIN_MAX_BLOCK_SIZE - probed_num;
                             let new_index = index + addition;
 
-                            build_indexes.extend_from_slice(&probed_rows[index..new_index]);
-                            probe_indexes.extend(repeat(i as u32).take(addition));
+                            for it in &probed_rows[index..new_index] {
+                                build_indexes.push(it);
+                            }
+                            probe_indexes[probe_indexes_len] = (i as u32, addition as u32);
+                            probe_indexes_len += 1;
+                            probed_num += addition;
 
                             probed_blocks.push(self.merge_eq_block(
-                                &self.row_space.gather(&build_indexes)?,
-                                &DataBlock::take(input, &probe_indexes)?,
+                                &self.row_space.gather_build(
+                                    &build_indexes,
+                                    &data_blocks,
+                                    &num_rows,
+                                )?,
+                                &DataBlock::take_by_compressd_indices(
+                                    input,
+                                    probe_indexes,
+                                    probe_indexes_len,
```

Comment:
Seems `probe_indexes_len` is same as `probe_indexes.len()`

Reply:
As we reuse the probe_indexes, therefore probe_indexes.len() is always equal to JOIN_MAX_BLOCK_SIZE.